_posts/2021-07-26-hhvm-4.120.markdown

```diff
@@ -0,0 +1,66 @@
+---
+title: "HHVM 4.120"
+layout: post
+author: fred
+category: blog
+---
+
+HHVM 4.120 is released! This release marks the end of support for 4.113 and
+4.114; HHVM 4.115&ndash;4.118 remain supported, as do the 4.80 and 4.102 LTS
+releases. HHVM 4.119 was skipped due to personal circumstances of the team.
+
+
+# Highlights
+
+- autocomplete can now provide documentation for suggestions
+- autocomplete now understands built-in primitive types like `string`
+- `stream_select()` and `pagelet_server_task_result()` will no longer be
+  inlined, so will appear correctly in profilers like Xenon.
+- added `HH\\Lib\\Legacy_FIXME` namespace, containing several functions emulating
+  PHP behavior that we intend to remove from the core language. New code should
+  never use these functions, and projects should aim to remove them over time.
+  Currently, these include:
+  - `Legacy_FIXME\\increment(mixed)` and `decrement(mixed)`: reimplements the
+    current behavior of `++` and `--` on `null` and `string` inputs.
+  - `Legacy_FIXME\\cast_for_arithmetic(mixed)`: match PHP's implicit coercion
+    for math operations on `null`, `bool`s, `string`s (int-like, float-like,
+    and neither), and `resource`s.
+  - `lt(mixed, mixed)`, `lte(mixed, mixed)`, `gt(mixed, mixed)`,
+    `gte(mixed, mixed)`, `cmp(mixed, mixed)`, `eq(mixed, mixed()`,
+    `neq(mixed, mixed)`
+  - `optional_arraykey_to_int_cast_for_switch(?arraykey $value)[]: int`
+  - `optional_num_to_int_cast_for_switch(?num $value)[]: int`
+  - `int_cast_for_switch(mixed $value, ?arraykey $first_truthy)`; the second
+    parameter is required when switching on an enum.
+  - `string_cast_for_basic_switch(mixed $value, ?arraykey $first_case)`
+
+These last few functions are intended to be used by replacing `switch($foo)`
+with `switch(Legacy_FIXME\\string_cast_for_basic_switch($foo))`, for example.
+
+
+# Breaking Changes
+
+- TODO: can not repro: 4ab60de666 'remove bitwise not for string in hackc"
```

Comment:
Still a TODO, not sure what's going on here. https://github.com/facebook/hhvm/commit/4ab60de666

Reply:
Probably going to publish this tomorrow so I know whether or not this is a breaking change in practice.

Reply:
This is a constant-folding change only, not observable.